educator/tests/Test/Dscp/DB/SQLite/Transactions.hs

```diff
@@ -0,0 +1,96 @@
+{-# LANGUAGE QuasiQuotes #-}
+
+-- | This module tests that our wrapper over SQLite library
+-- allow SQL transactions to work properly.
+-- Transactions do not work as expected automatically; for instances,
+-- queries are not allowed to be performed concurrently via the same connection.
+module Test.Dscp.DB.SQLite.Transactions where
+
+import qualified Control.Concurrent.STM as STM
+import qualified Data.List as L
+import Database.SQLite.Simple (Only (..))
+import Loot.Base.HasLens (HasCtx)
+import System.Directory (removeFile)
+import Text.InterpolatedString.Perl6 (q)
+import UnliftIO (MonadUnliftIO)
+import qualified UnliftIO.Async as UIO
+
+import Dscp.DB.SQLite
+import Dscp.Rio
+import Dscp.Util.Test
+
+type MonadMoney m = (MonadIO m, MonadCatch m, MonadUnliftIO m)
+
+type Money = Int
+
+prepareSchema :: (MonadIO m) => DBT t 'Writing m ()
+prepareSchema =
+    forM_ [createTableQuery, addAccountQuery] $
+        \que -> execute que ()
+  where
+    createTableQuery = [q|
+        create table if not exists Accounts (
+            amount    INTEGER
+        );
+        |]
+    addAccountQuery = [q|
+        insert into Accounts values (0);
+        |]
+
+getMoney :: MonadIO m => DBT t w m Money
+getMoney = fromOnly . L.head <$> query queryText ()
+  where
+    queryText = [q|
+        select amount from Accounts
+    |]
+
+setMoney :: MonadIO m => Money -> DBT t 'Writing m ()
+setMoney val = execute queryText (Only val)
+  where
+    queryText = [q|
+        update Accounts set amount = ?
+    |]
+
+addMoney :: (MonadUnliftIO m, HasCtx ctx m '[SQLiteDB]) => m ()
+addMoney =
+    transactW @'WithinTx $ do
+        money <- getMoney
+        setMoney (money + 1)
+
+runSQLiteMode :: RIO SQLiteDB a -> IO a
+runSQLiteMode action = do
+    db' <- openSQLiteDB dbParams
+    runRIO db' $ invoke prepareSchema
+    closeSQLiteDB db'
+    bracket (openSQLiteDB dbParams)
+            (\db -> closeSQLiteDB db >> removeFile dbPath) $
+            \db -> runRIO db action
+  where
+    -- sad, but testing with in-memory database is not an option
+    -- because each connection would work with its own database in this case
+    dbPath = "./sql-transaction-test.db"
```

Comment:
This is so sad. Did you try to use single connection with WAL? Any chance it will work properly?

Reply:
Right, let me try

Reply:
No, it doesn't :pepe_sad: